### PR TITLE
Prevent reading broken IR caches by changing their serialVersionUID

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Many engine PRs modify builtins or other engine internals and then they are subject to [incremental CI runtime errors](https://www.pivotaltracker.com/n/projects/2539304/stories/182868680) as outdated `IR` caches from global space at `~/.local/share/enso/cache/ir/Standard/Builtins/0.0.0-dev/0.0.0-dev/Standard/Builtins` are read in.

This PR provides solution for that problem by explicitly defining `IR.Module` `serialVersionUID`. By changing the `serialVersionUID` one prevents previously saved `IR` caches to be loaded into the running process. Change the `serialVersionUID` whenever you see errors caused by reading outdated `IR` caches in the CI.

### Important Notes

Whenever one needs to avoid loading previous `IR` caches, go to `case class IR.Module` and change the `@SerialVersionUID(3692L)` to **number of your pull request**.

### Checklist

Please include the following checklist in your PR:

- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md) style guides.
- All code has been tested:
  - it has been verified in debugger that changing the `serialVersionUID` prevents reading previous `IR` caches

[ci no changelog needed]

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala [hunk 1]
```diff
@@ -298,6 +298,7 @@ object IR {
     * @param passData the pass metadata associated with this node
     * @param diagnostics compiler diagnostics for this node
     */
+  @SerialVersionUID(3692L) // prevents reading broken caches, see PR-3692 for details
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: When CI fails due to errors caused by reading previous `IR` caches, change the `3692L` value to a different unique value. Number of your PR should be unique enough.

Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: In my understanding the caching logic was:
- try to deserialize the IR
- if it fails, compile and re-populate the cache

You mean that here are cases when this approach doesn't work?

**Comment by mwu-tow** [comment]
@JaroslavTulach 
The CI script currently clears `~/.local/share/enso/cache` before building engine. Or at least it is supposed to.

Is this PR meant to remove the need for such clearing or are there issues with removing this cache? The latter would be a serious concern for me.

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: There are many examples where it didn't work. Whole week I was fighting with failing CI: https://github.com/enso-org/enso/runs/8190285398?check_suite_focus=true
```
INFO ide_ci::program::command: sbtℹ️ [error] Test org.enso.interpreter.test.VectorTest.passingListOrArrayToEnsoAsArray failed: org.graalvm.polyglot.PolyglotException: org.enso.compiler.exception.CompilerError: Compiler Internal Error: Constructor for Cons(Vector_Data,1,false) is null, took 5.316 sec
 INFO ide_ci::program::command: sbtℹ️ [error]     at org.enso.compiler.codegen.IrToTruffle$ExpressionProcessor.processName(IrToTruffle.scala:1190)
 INFO ide_ci::program::command: sbtℹ️ [error]     at org.enso.compiler.codegen.IrToTruffle$ExpressionProcessor.run(IrToTruffle.scala:722)
 INFO ide_ci::program::command: sbtℹ️ [error]     at org.enso.compiler.codegen.IrToTruffle$ExpressionProcessor.run(IrToTruffle.scala:715)
```
we need a way to tell that the caches are out of date, especially when they are stored in a shared location.

By specifying different `SerialVersionUID` a developer (like us) can explicitly tell the system that format/content of caches is now  different. Another approach would be to ignore all caches that are older than the `runtime.jar` itself. One can get access to `runtime.jar` file via `MyClass.class.getProtectionDomain().getCodeSource().getLocation()`; we could compare the `.ir` file timestamp and ignore reading the outdated ones...

**Comment by JaroslavTulach** [comment]
> @JaroslavTulach The CI script currently clears `~/.local/share/enso/cache` before building engine. Or at least it is supposed to.
> 
> Is this PR meant to remove the need for such clearing or are there issues with removing this cache? The latter would be a serious concern for me.

There are serious issues caused by existence of such cache! Especially when two different (development) versions share the same cache, we can get into inconsistent situations like https://github.com/enso-org/enso/runs/8190285398?check_suite_focus=true - with the cache format versioning we can at least detect such situation and when [CI fails](https://github.com/enso-org/enso/pull/3692#discussion_r965621351) change the `serialVersionUID`.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Improved Vector/Array interop · enso-org/enso@9ab4c64</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by mwu-tow** [comment]
> There are serious issues caused by existence of such cache! Especially when two different (development) versions share the same cache, we can get into inconsistent situations like https://github.com/enso-org/enso/runs/8190285398?check_suite_focus=true - with the cache format versioning we can at least detect such situation and when [CI fails](https://github.com/enso-org/enso/pull/3692#discussion_r965621351) change the `serialVersionUID`.

@JaroslavTulach 
I understand. 
My point is that CI has already been deleting the cache before building engine. There's literally a log line about removing `/home/mwu/.local/share/enso/cache`. This means that either:
* the linked build failure is caused by other artifacts/cache paths — then the description of this PR would be wrong;
* the CI somehow fails to properly clear the cache — then I must have broken something terribly and I should immediately investigate.

Are there any other possible reasons aside from the global cache to cause such problem?

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Improved Vector/Array interop · enso-org/enso@9ab4c64</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by JaroslavTulach** [comment]
> > There are serious issues caused by existence of such cache! Especially when two different (development) versions share the same cache, we can get into inconsistent situations like https://github.com/enso-org/enso/runs/8190285398?check_suite_focus=true - with the cache format versioning we can at least detect such situation and when [CI fails](https://github.com/enso-org/enso/pull/3692#discussion_r965621351) change the `serialVersionUID`.
> 
> @JaroslavTulach I understand. My point is that CI has already been deleting the cache before building engine. There's literally a log line about removing `/home/mwu/.local/share/enso/cache`. This means that either:
> 
>     * the linked build failure is caused by other artifacts/cache paths — then the description of this PR would be wrong;
> 
>     * the CI somehow fails to properly clear the cache — then I must have broken something terribly and I should immediately investigate.
> 
> 
> Are there any other possible reasons aside from the global cache to cause such problem?

I am constantly fighting with CI in #3667 - yesterday I executed the [clean actions](https://github.com/enso-org/ci/actions/workflows/clean.yml) and it did help temporarily, but today I see the old failures caused by outdated caches again. Examples: 
- https://github.com/enso-org/enso/runs/8252188214?check_suite_focus=true
- https://github.com/enso-org/enso/runs/8252187942?check_suite_focus=true

Anyway this PR isn't about CI. I can reproduce the same problem locally. This PR just gives developers a way to clearly communicate a change in the IR caches format.


<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Improved Vector/Array interop · enso-org/enso@9ab4c64</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Improved Vector/Array interop · enso-org/enso@ef71c20</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Improved Vector/Array interop · enso-org/enso@ef71c20</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Review by 4e6 — APPROVED** [review]
Let's try

**Pull request merged**